pkg/cache/application_info.go

```diff
@@ -112,6 +127,23 @@ func (ai *ApplicationInfo) IsWaiting() bool {
 // The state machine handles the locking.
 func (ai *ApplicationInfo) HandleApplicationEvent(event applicationEvent) error {
 	err := ai.stateMachine.Event(event.String(), ai)
+	updatedApps := make([]*si.UpdatedApplication, 0)
+	updatedApps = append(updatedApps, &si.UpdatedApplication{
+		ApplicationID: ai.ApplicationID,
+		State:         ai.stateMachine.Current(),
+		StateTransitionTimestamp: time.Now().UnixNano(),
+		Message: fmt.Sprintf("{Status change triggered by the event : %s}",event.String()),
+	})
+
+	if err ==nil && ai.statusChangeEventHandler != nil {
+	ai.statusChangeEventHandler.HandleEvent(
+		&rmevent.RMApplicationUpdateEvent{
+			RmID:                 ai.rmId,
+			AcceptedApplications: make([]*si.AcceptedApplication, 0),
+			RejectedApplications: make([]*si.RejectedApplication, 0),
+			UpdatedApplications: updatedApps,
+		})
+	}
```

Comment:
do not handle this here, we can add this to the callback while state changes. https://github.com/apache/incubator-yunikorn-core/blob/a4a44d7e05d16aa712e1777c85f09e4d55466830/pkg/cache/application_state.go#L102-L108.

This is called whenever there is a state change
> // 6. enter_state - called after entering all states

Like the rest of callbacks in `application_state.go`, you can create a function in `application_info.go`, such as

```
func (ai *ApplicationInfo) onStateChange() {
  log.Logger().Debug("Application state transition",
    zap.String("appID", event.Args[0].(*ApplicationInfo).ApplicationID),
    zap.String("source", event.Src),
    zap.String("destination", event.Dst),
    zap.String("event", event.Event))

  // TODO : move the event handling here
} 
```

Reply:
> Hi @kingamarton
> please see my comments inline. Also could you please add some UT for this as well?

@yangwwei do you have any good idea on how can we test in from the unit test (without writing integration test)? I was thinking to write some test for the onStatusChange method, but I don't have any good idea what can we test there. THis is something what we can easily test in out integration tests, but for unit tests if you have any good ideas, please let me know.

Reply:
@kingamarton I think you can do something similar to https://github.com/apache/incubator-yunikorn-core/blob/master/pkg/cache/application_state_test.go. You can create some app with `NewApplicationInfoWithEventHandler `, then a mocked implementation of the eventHandler, then verify the state info is actually handled.